Big Mode for OpenSlides

ostcar opened the pull request:
Uses django channels instead of tornado for the autoupdate. Therefore
tornado is nolonger a dependency of OpenSlides (but channels).

This uses websockets instead of sockJS. All modern browsers should
support websockets.

This commit is still work in progress.
TODOS:
- [x] The config app has a rase condition in the small mode, fixed with #2175 
- [x] I patched the channel app. So we have to wait for an update on channels or find another way around my problem.
- [x] Create an ```OpenSlidesSessionBackend``` that saved the user id directly to the db (or the cache?) to not all sessions have to be decoded on any autoupdate.
- [x] Fix README, some changed done, for the rest see #2177

@normanjaeckel can you look at this and help me with the config problem?

ostcar commented:
@normanjaeckel This is ready for review. If it runs in the portable, then it can be merged.

normanjaeckel commented on CHANGELOG:
```diff
@@ -4,6 +4,13 @@
+- Use channels instead of tornado.
+- Therefore support big assemblies, with lots of users.
+
 Version 2.0.1 (unreleased)
```
You can merge the list for 2.0.1 into the new list for 2.1 because we won't release a 2.0.1 anymore if the channels thing is in master.

ostcar replied:
done

normanjaeckel commented on CHANGELOG:
```diff
@@ -4,6 +4,13 @@
+Version 2.1 (unreleased)
+========================
+
+Core:
```
Proposal:
```
Core:
- Used Django Channels instead of Tornado.
- Added support for big assemblies with lots of users.
```

ostcar replied:
done

normanjaeckel commented on README.rst:
```diff
@@ -83,7 +83,7 @@ If you run this command the first time, a new database and the admin
 change the password after first login!
 
-OpenSlides will start using the integrated Tornado webserver. It will also
+OpenSlides will start a webserver. It will also
```
Proposal: Fit all lines of this paragraph to the default line ending (`Ctrl+J` in Geany).

ostcar replied:
done (in atom it is ctrl+alt+q)

normanjaeckel commented on README.rst:
```diff
@@ -122,6 +122,29 @@ If you want to contribute to OpenSlides, have a look at `OpenSlides website
+* https://github.com/sebleier/django-redis-cache
+* https://docs.djangoproject.com/en/1.9/ref/settings/#databases
+
+You should use a webserver like apache or nginx to serve the static files and
```
Proposal: `You should use a webserver like Apache HTTP Server or nginx to serve the static and media files and as proxy server in front of OpenSlides' server. You also should use a database like PostgreSQL.`

ostcar replied:
done (with a hint to redis)

normanjaeckel commented on README.rst:
```diff
@@ -140,6 +163,8 @@ OpenSlides uses the following projects or parts of them:
 
 * `html5lib <https://github.com/html5lib/html5lib-python>`_, License: MIT
 
+* `channels <https://github.com/andrewgodwin/channels/>`_, License: MIT
```
Proposal:
```
* `Django Channels <https://github.com/andrewgodwin/channels/>`_, License: MIT
```

ostcar replied:
The project seems to be called only channels. So would keep it. Or do you have any source, where it is called "Django Channels"? I don't know if channels can be used for other APIs like flask.

normanjaeckel replied:
https://github.com/andrewgodwin/channels/blob/master/README.rst

ostcar replied:
ok, done

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -252,3 +253,19 @@ class Meta:
+
+    @classmethod
+    def get_session_store_class(cls):
+        from .session import SessionStore
```
This should raise an Import Error.

ostcar replied:
done

normanjaeckel commented on openslides/global_settings.py:
```diff
@@ -8,6 +8,8 @@
 AUTHENTICATION_BACKENDS = ('openslides.users.auth.CustomizedModelBackend',)
 
+# Uses a db session backend, that saves the user_id directly in the db
+SESSION_ENGINE = 'openslides.core.sessions'
```
Proposal: Rename the module to `session_backend`.

ostcar replied:
done

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -1,165 +1,26 @@
+def get_active_users():
     """
+    Helper to get all logged in users. Also returns the anonymous user if there
+    is at least one session with it.
```
The code looks like that the anonymous is always present ...

ostcar replied:
fixed by optimizing the query. Also renamed the function, because now it does never return the AnonymousUser

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -170,31 +31,32 @@ def inform_changed_data(is_delete, *args):
+            if action == 'changed':
+                data = access_permissions.get_restricted_data(full_data, user)
+                if data is None:
+                    # There are no data for the user so he can't see the object. Skip him.
```
Hm. Maybe we nevertheless want to inform him that the object he can't see has changed.

ostcar replied:
That is your old code with your old comment. I would like to keep it, because I don't now how the client would react of a change and I don't think we should change it.

ostcar commented:
Work in progress. Please wait.

ostcar commented:
Ready for review and merge

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -1,211 +1,132 @@
+                'is_deleted': is_deleted and instance == root_instance,
+                'dispatch_uid': root_instance.get_access_permissions().get_dispatch_uid()})
+        except ChannelLayer.ChannelFull:
+            pass
```
Please add a warning if the start command is used.

emanuelschuetze commented:
I tested your current branch:
* On Linux: I get a 404 for js-data-http.js.map: ` HTTP GET /static/js/js-data-http.js.map 404`
* Windows portable works now.

emanuelschuetze commented:
One bug was introduces with 2221b23447bc996fc8f5f073a6cd0e868e6b4028 (Redesign of the config variables.) which should fix the big mode:
After db creation all config values are always in English (also if my browser language is German). See e.g. title and text of welcome page after new db creation. Before this commit the config value language depends on the browser language (correct).

normanjaeckel commented on openslides/asgi.py:
```diff
@@ -0,0 +1,6 @@
+import os
+from channels.asgi import get_channel_layer
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "my_project.settings")
```
Please use setup_django_settings_module().

ostcar commented:
@normanjaeckel I fixed your remarks and also changed the default setting. Please review once more.

@emanuelschuetze I did not change anything at the client that could explain you 404. Please add an issue for that. I did not understand what you mean with your second comment.

normanjaeckel commented:
Ok. Rebase and merge.

emanuelschuetze commented:
@ostcar Does you need the file js-data-http.js.map? Seems that gulp this file does not copy to static...

ostcar commented:
@emanuelschuetze I don't need this file (yet) but this is not a problem of this pull request. Please create an issue if this problem does still happen.
